iModelCore/ECPresentation/Source/Content/ContentSpecificationsHandler.cpp

```diff
@@ -367,6 +367,55 @@ bvector<std::unique_ptr<FlattenedRelatedPropertiesSpecification>> FlattenedRelat
     return specs;
     }
 
+/*---------------------------------------------------------------------------------**//**
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+static bool HasSamePath(RelatedClassPath const& relatedClassPath, bvector<RelationshipStepSpecification*> const& specsPath, ECSchemaHelper const& helper)
+    {
+    if (relatedClassPath.size() != specsPath.size())
+        return false;
+
+    bool allStepsMatch = true;
+    for (size_t i = 0; i < specsPath.size(); ++i)
+        {
+        auto const& createdStepSpec = specsPath[i];
+        auto const& includeStep = relatedClassPath[i];
+
+        ECClassCP specStepsRelationshipClass = helper.GetECClass(createdStepSpec->GetRelationshipClassName().c_str());
+        bool hasSameRelationshipClass = includeStep.GetRelationship().GetClass().Is(specStepsRelationshipClass);
+        bool hasSameDirection = includeStep.IsForwardRelationship() == (createdStepSpec->GetRelationDirection() == RequiredRelationDirection_Forward);
+        if (!hasSameRelationshipClass || !hasSameDirection)
+            {
+            allStepsMatch = false;
+            break;
+            }
+        }
+    return allStepsMatch;
+    }
+
+/*---------------------------------------------------------------------------------**//**
+* @bsimethod
++---------------+---------------+---------------+---------------+---------------+------*/
+static bvector<std::unique_ptr<FlattenedRelatedPropertiesSpecification>> FilterFlatSpecsByExclusiveIncludePaths(bvector<std::unique_ptr<FlattenedRelatedPropertiesSpecification>>& flatSpecs,
+    std::shared_ptr<RelatedClassPathsList> const& exclusiveIncludePaths, ECSchemaHelper const& helper)
+    {
+    bvector<std::unique_ptr<FlattenedRelatedPropertiesSpecification>> filteredFlatSpecs;
+    for (auto& spec : flatSpecs)
+        {
+        for (auto const& relatedClassPath : *exclusiveIncludePaths)
+            {
+            if (HasSamePath(relatedClassPath, spec->GetFlattened().GetPropertiesSource()->GetSteps(), helper))
+                {
+                std::unique_ptr<FlattenedRelatedPropertiesSpecification> dummy;
+                std::swap(spec, dummy);
+                filteredFlatSpecs.push_back(std::move(dummy));

```

Comment:
It's really non-obvious that the input vector is modified by calling this. Could we change this function to one of these:
- return void and modify the input in-place
- return vector, but take input as const

?